File: backend.scala/app/com/skewwhiffy/auraltester/notes/Interval.scala
package com.skewwhiffy.auraltester.notes

import com.skewwhiffy.auraltester.notes.Interval.{displayStrings, perfectDegrees}

object Interval {
  lazy val displayStrings: List[String] = List(
    "unison",
    "second",
    "third",
    "fourth",
    "fifth",
    "sixth",
    "seventh",
    "octave"
  )

  lazy val perfectDegrees: Set[Int] = Set(1, 4, 5, 8)

  lazy val augmented: Int => Interval = degree => {
    val baseInterval = if (perfectDegrees.contains(degree)) perfect(degree) else major(degree)
    baseInterval.augmented
  }

  lazy val diminished: Int => Interval = degree => {
    val baseInterval = if (perfectDegrees.contains(degree)) perfect(degree) else minor(degree)
    baseInterval.diminished
  }

  lazy val minor: Int => Interval = degree => major(degree).diminished

  lazy val major: Int => Interval = degree => {
    if (perfectDegrees.contains(degree)) {
      throw new IllegalArgumentException(s"Cannot instantiate major interval of degree '$degree'")
    }
    Interval(degree, 0)
  }

  lazy val perfect: Int => Interval = degree => {
    if (!perfectDegrees.contains(degree)) throw new IllegalArgumentException(s"Cannot instantiate perfect interval of degree '$degree'")
    else Interval(degree, 0)
  }
}

case class Interval(degree: Int, deviation: Int) {
  lazy val diminished: Interval = Interval(degree, deviation - 1)

  lazy val augmented: Interval = Interval(degree, deviation + 1)

  lazy val displayString: String = s"$quality ${displayStrings(degree - 1)}"

  lazy val up: DirectedInterval = DirectedInterval(this, IntervalDirection.up)

  lazy val down: DirectedInterval = DirectedInterval(this, IntervalDirection.down)

  private def quality: String = {
    lazy val defaultQuality = if (perfectDegrees.contains(degree)) "perfect" else "major"
    lazy val negativeQuality = {
      if (!perfectDegrees.contains(degree) && deviation == -1) "minor"
      else {
        (if (perfectDegrees.contains(degree)) -deviation else -deviation - 1) match {
          case 1 => "diminished"
          case 2 => "doubly diminished"
          case it => s"${it}x diminished"
        }
      }
    }
    lazy val positiveQuality = (deviation match {
      case 1 => ""
      case 2 => "doubly "
      case it => s"${it}x "
    }) + "augmented"

    deviation match {
      case 0 => defaultQuality
      case it if it < 0 => negativeQuality
      case it if it > 0 => positiveQuality
    }
  }

  override def equals(obj: Any): Boolean = obj match {
    case other: Interval => other.deviation == deviation && other.degree == degree
    case _ => false
  }

  override def toString: String = displayString
}

File: backend/src/main/kotlin/com/skewwhiffy/notation/model/interval/Interval.kt
package com.skewwhiffy.notation.model.interval

data class Interval(val degree: Int, val deviation: Int) {
  companion object {
    private val displayStrings: List<String> = listOf(
      "unison",
      "second",
      "third",
      "fourth",
      "fifth",
      "sixth",
      "seventh",
      "octave"
    )

    private val perfectDegrees: Set<Int> = setOf(1, 4, 5, 8)

    fun augmented(degree: Int): Interval {
      val baseInterval = if (perfectDegrees.contains(degree)) perfect(degree) else major(degree)
      return baseInterval.augmented
    }

    fun diminished(degree: Int): Interval {
      val baseInterval = if (perfectDegrees.contains(degree)) perfect(degree) else minor(degree)
      return baseInterval.diminished
    }

    fun minor(degree: Int): Interval = major(degree).diminished

    fun major(degree: Int): Interval {
      if (perfectDegrees.contains(degree)) throw IllegalArgumentException("Cannot instantiate major interval of degree '$degree'")
      return Interval(degree, 0)
    }

    fun perfect(degree: Int): Interval {
      if (!perfectDegrees.contains(degree)) throw IllegalArgumentException("Cannot instantiate perfect interval of degree '$degree'")
      return Interval(degree, 0)
    }
  }

  val diminished: Interval get() = Interval(degree, deviation - 1)

  val augmented: Interval get() = Interval(degree, deviation + 1)

  val displayString: String get() = "$quality ${displayStrings[degree - 1]}"

  val up: DirectedInterval = DirectedInterval(this, IntervalDirection.UP)

  val down: DirectedInterval = DirectedInterval(this, IntervalDirection.DOWN)

  private val quality: String
    get() {
      val defaultQuality = if (perfectDegrees.contains(degree)) "perfect" else "major"
      val negativeQuality =
        if (!perfectDegrees.contains(degree) && deviation == -1) "minor"
        else (if (perfectDegrees.contains(degree)) -deviation else -deviation - 1).let {
          when (it) {
            1 -> "diminished"
            2 -> "doubly diminished"
            else -> "${it}x diminished"
          }
        }

      val positiveQuality = when (deviation) {
        1 -> ""
        2 -> "doubly "
        else -> "${deviation}x "
      } + "augmented"

      return when {
        deviation < 0 -> negativeQuality
        deviation > 0 -> positiveQuality
        else -> defaultQuality
      }
    }

  override fun toString(): String = displayString
}
